packages/verkle/src/db/checkpoint.ts

```diff
@@ -0,0 +1,275 @@
+    this.checkpoints = []
+
+    if (this.cacheSize > 0) {
+      // @ts-ignore
+      this._cache = new LRUCache({
+        max: this.cacheSize,
+        updateAgeOnGet: true,
+      })
+    }
+  }
+
+  /**
+   * Flush the checkpoints and use the given checkpoints instead.
+   * @param {Checkpoint[]} checkpoints
+   */
+  setCheckpoints(checkpoints: Checkpoint[]) {
+    this.checkpoints = []
+
+    for (let i = 0; i < checkpoints.length; i++) {
+      this.checkpoints.push({
+        root: checkpoints[i].root,
+        keyValueMap: new Map(checkpoints[i].keyValueMap),
+      })
+    }
+  }
+
+  /**
+   * Is the DB during a checkpoint phase?
+   */
+  hasCheckpoints() {
+    return this.checkpoints.length > 0
+  }
+
+  /**
+   * Adds a new checkpoint to the stack
+   * @param root
+   */
+  checkpoint(root: Uint8Array) {
+    this.checkpoints.push({ keyValueMap: new Map<string, Uint8Array>(), root })
+  }
+
+  /**
+   * Commits the latest checkpoint
+   */
+  async commit() {
+    const { keyValueMap } = this.checkpoints.pop()!
+    if (!this.hasCheckpoints()) {
+      // This was the final checkpoint, we should now commit and flush everything to disk
+      const batchOp: BatchDBOp[] = []
+      for (const [key, value] of keyValueMap.entries()) {
+        if (value === undefined) {
+          batchOp.push({
+            type: 'del',
+            key: unprefixedHexToBytes(key),
+          })
+        } else {
+          batchOp.push({
+            type: 'put',
+            key: unprefixedHexToBytes(key),
+            value,
+          })
+        }
+      }
+      await this.batch(batchOp)
+    } else {
+      // dump everything into the current (higher level) diff cache
+      const currentKeyValueMap = this.checkpoints[this.checkpoints.length - 1].keyValueMap
+      for (const [key, value] of keyValueMap.entries()) {
+        currentKeyValueMap.set(key, value)
+      }
+    }
+  }
+
+  /**
+   * Reverts the latest checkpoint
+   */
+  async revert() {
+    const { root } = this.checkpoints.pop()!
+    return root
+  }
+
+  /**
+   * @inheritDoc
+   */
+  async get(key: Uint8Array): Promise<Uint8Array | undefined> {
+    const keyHex = bytesToUnprefixedHex(key)
+    if (this._cache !== undefined) {
+      const value = this._cache.get(keyHex)
+      this._stats.cache.reads += 1
+      if (value !== undefined) {
+        this._stats.cache.hits += 1
+        return value
+      }
+    }
+
+    // Lookup the value in our diff cache. We return the latest checkpointed value (which should be the value on disk)
+    for (let index = this.checkpoints.length - 1; index >= 0; index--) {
+      if (this.checkpoints[index].keyValueMap.has(keyHex)) {
+        return this.checkpoints[index].keyValueMap.get(keyHex)
+      }
+    }
+    // Nothing has been found in diff cache, look up from disk
+    const valueHex = await this.db.get(keyHex, {
+      keyEncoding: KeyEncoding.String,
+      valueEncoding: ValueEncoding.String,
```

Comment:
I think with Verkle we use a completely new database, right? (So not related to old MPT DB). If that's the case, can we immediately use `valueEncoding` set to `ValueEncoding.Bytes`? Then we don't have to convert `bytesToUnprefixedHex` early on and we don't have to deal with a breaking release for the underlying database later on.

Reply:
(same maybe also for keys directly)

Reply:
Good point, will go with bytes.

Reply:
Updated

Reply:
I've updated all key/value types to Uint8Arrays (db, cache), but kept the Checkpoint `keyValueMap` as `Map<string, Uint8Array | undefined>` since Uint8Arrays can't directly be used as Map keys (since two uint8arrays are not equal even if they contain the same data)

Reply:
I'm now reading that using Uint8Array as key for the cache (the cache uses the LRUCache package) might cause issues. I will do some testing but I think I'll revert that part, so we'd end up with:

- A DB that uses Uint8Arrays for keys and values
- A cache (LRU-cache) that uses `string` for keys and Uint8Array as values
- A keyValueMap (`Checkpoint` type) that uses `string` as keys and Uint8Array as values.

Reply:
Indeed, this fails:

![image](https://github.com/ethereumjs/ethereumjs-monorepo/assets/18757482/e525673c-c76e-41b9-8f66-c3e7708d948b)

So I'll switch the cache to use hexstrings for keys.